Potsanity 3.0

rrealmuto opened the pull request:
This is still a WIP and needs more testing but wanted to put it on people's radar.

Stuff the player will notice:
- Shuffle empty pots/crates and fairy pots
This PR adds previously unshuffled pots and crates to the location pool. These are pots that are either empty or contain a fairy. Empty pots/crates have a "Nothing" item shuffled into the item pool, which, as the name suggests, does literally nothing. Fairy pots introduce a "stray fairy" item to the pool, which gives a full heal when collected. 

![GLideN64_THE_LEGEND_OF_ZELDA_019](https://github.com/OoTRandomizer/OoT-Randomizer/assets/5086838/981baeae-16a6-4548-a593-021cbf265a04)

![GLideN64_THE_LEGEND_OF_ZELDA_020](https://github.com/OoTRandomizer/OoT-Randomizer/assets/5086838/b64f2227-8f6d-4ee3-a6ee-0cd4ff07792b)

- Enhanced pot major item effect
When collecting a shuffled major item from freestanding/pot/crate/etc. checks, it used to turn invisible upon collecting it. Now, the item will display above links head until the message box is closed.

https://github.com/OoTRandomizer/OoT-Randomizer/assets/5086838/25ccdba4-f4cb-4e65-a79a-3b226070ebd0


The under the hood stuff:
This PR completely reworks the custom actor flag system that was introduced with potsanity. 

- Flag storage within an actor/Extending Actor struct:
The old system worked by storing unique flags for each actor into sort-of unused home rotation variables within the actor instance. This new system adds a hack that effectively increases the base Actor structure in order to store this flag information. While only being used for a handful of actors currently, this hack in an enabler for pretty much any other shuffle setting imaginable. 
- Flag info
Flags were previously a unique 16 bit identifier which consisted of the scene setup, room, and actor index for each actor. This PR extends the flag to 32 bits in order to include the scene index, and more importantly a "sub-flag". This is used by certain actors which spawn multiple shuffled items (goron pot, rupee towers, etc.). Those items previously had to be handled as special cases within the patcher. Now they are treated the same as any other shuffled item utilizing the new flags.
- Flag tables/SRAM storage
This PR also changes how these flags are stored in SRAM to be significantly more compact. Previously, a large amount of the SRAM used by these new flags were unused bits which represented actors within a room that weren't shuffled. This new system ensures that there are no wasted bits stored in SRAM (at the expense of storing some more data in tables within the code payload).

fenhl commented:
The changes to override keys will require changes to the co-op context.

r0bd0g commented:
Here's a funny logic thing I thought of just now that I should probably post in case I forget about it later.

I assume you have to collect the item in the pot before the fairy will spawn next time? This causes a problem with the logic of the fairy pot in the bk chest room of MQ Fire, which previously expected you could obtain the fairy without having physical access to the pot. So you'll actually have to check whether that pot is shuffled, and if so, hard require the hookshot. In addition to needing to check whether dungeon pots are shuffled, because of the way pots revert to vanilla behaviour when disabled, this means the logic will need to be able to check whether the location is disabled, which I'm not sure is currently possible.

TreZc0 commented:
Reminder for maintainers to heavily test encryption with this once it has been merged 🙂


Great job!

fenhl commented:
> So I'm assuming that on the game's side, all that needs to happen is incrementing the COOP_VERSION and increasing the size of OUTGOING_KEY?

I would also recommend moving `OUTGOING_KEY` to `0x0C1C` instead of shifting everything else. Whatever changes are made will also need to be documented at `Notes/coop-ctx.md`.

> Something will need to happen on the multiworld client side too, right?

Yes, each individual multiworld client will have to add support for the new co-op context version. In the past, the maintainers have required sending a corresponding PR to [bizhawk-co-op](https://github.com/TestRunnerSRL/bizhawk-co-op) to add minimal support for the new co-op context version, not sure if that's still the case.

fenhl commented on ASM/c/get_items.c:
```diff
@@ -284,14 +285,12 @@ void move_outgoing_queue() {
         }
-        if (z64_file.scene_flags[key_scene].unk_00_ == override.key.all) {
-            // Prevent duplicate entries
+        if (extended_savectx.incoming_queue[i].key.all == override.key.all) {
```
Keep the comment.

fenhl commented on ASM/src/coop_state.asm:
```diff
@@ -23,6 +23,7 @@ MW_PROGRESSIVE_ITEMS_ENABLE:
 
 OUTGOING_KEY:
 .word 0x00000000
+.word 0x00000000
```
To make it easier to support multiple versions of the co-op context at once, replace this with a 4-byte gap and move the `OUTGOING_KEY` symbol to the end of the co-op context. This preserves the relative addresses of the other symbols.

fenhl reviewed (commented):
The changes to the co-op context need to be documented in `Notes/coop-ctx.md`, and the changes to the save data in `Notes/extra-save-data.md` (might be a good opportunity to start documenting the contents of the extended save context).

r0bd0g commented:
I have some reservations about which crates are shuffled in which times of day. I think also some changes are probably needed regarding under what conditions the shield pot is shuffled, and what item it adds to the item pool. None of that really relates to the logic though. I have a couple of ideas for tricks but they're a bit weird b/c of how the drops tend move a bit randomly, so I'll just jot them down and maybe handle them another time.


Logic Review:

1. The crates at the top of vanilla Fire are behind a bombable wall, so explosives are also required. (In truth, the hammer alone can be used, but you have to reload the room after hammering the fake door, since the door and the bombable wall are tied to the same flag. But you can see with the logic for the hammer chest, that we have opted not to include that in logic, and even if we did, it would have to be a trick.)

2. The logic for fairy pot in the BK Chest Room of MQ Fire relies on destroying the pot from afar using a charged magic spin, and then waiting for the fairy to fly over to you (which it always does). If the pot is shuffled, you'll have to collect the drop first, which means the Hookshot will be required to get over there and pick it up. So the logic for the fairy drop needs to check that either the relevant settings for shuffling the fairy pot are not enabled, or that you have the Hookshot. The way that pots work when they are disabled is different from how other checks work -- they revert fully to vanilla behaviour, rather than having the check guaranteed to be a useless item. So you can additionally check for the shuffled fairy pot location being disabled as another alternative to Hookshot. We might not currently have a means of checking whether a location is disabled inside the logic files. It's kind of up to you whether you want to actually implement this, or maybe just leave a comment explaining the situation.

3. The pots near the boss room of vanilla Water shouldn't check for Longshot, as Longshot is not required if you were to enter the dungeon from the boss door. The checks are already in a region that checks for Longshot when coming into it from the rest of the dungeon.

fenhl commented on ASM/c/item_table.c:
```diff
@@ -328,6 +328,8 @@ item_row_t item_table[] = {
     [0x0117] = ITEM_ROW(0x53,      SILVER_CHEST, 0x41, -1, 0x00A1, 0x01A1, 0x88, no_upgrade, give_small_key, CASTLE_ID, -1, resolve_text_small_keys), // Ganon's Castle Small Key
     [0x0118] = ITEM_ROW(0x53,      SILVER_CHEST, 0x41, -1, 0x913E, 0x01A2, 0x89, no_upgrade, give_small_key, TCG_ID,    -1, resolve_text_small_keys_cmg), // Small Key (Chest Game)
 
+    [0x0119] = ITEM_ROW(0x53, HEART_CHEST_SMALL, 0x41, -1, 0x90B4, 0x0177, 0x95, no_upgrade, full_heal, -1, -1, NULL), // Fairy
```
```suggestion
    [0x0119] = ITEM_ROW(0x53,       BROWN_CHEST, 0x41, -1, 0x90B4, 0x0177, 0x95, no_upgrade, full_heal, -1, -1, NULL), // Fairy
```

Heart chests are for items that matter for heart wincons. Recovery hearts and double defence don't use them so fairies shouldn't either.

fenhl commented on ASM/src/build.asm:
```diff
@@ -142,6 +140,8 @@ TRIFORCE_ICON_TEXTURE:
 
 .align 0x10
 
+.skip 0x200 ; Temporary address bump to avoid audio issues
```
![A Minecraft screenshot showing a simple bridge crossing a narrow river. In front of the bridge, there are signs reading “Temporary Bridge (don't judge)”, “Temporary means Permanent (not judging)”, and “h*ck”](https://github.com/OoTRandomizer/OoT-Randomizer/assets/641386/b7b08747-8b24-42cf-b7f5-b9dec83023f2)

fenhl commented on LocationList.py:
```diff
@@ -1113,7 +1174,7 @@ def shop_address(shop_id: int, shelf_id: int) -> int:
     ("Bottom of the Well MQ Center Room Right Pot 2",                ("Pot",          0x08,  (0,0,43), None,                        'Arrows (10)',                           ("Bottom of the Well MQ", "Master Quest", "Pots",))),
     ("Bottom of the Well MQ Center Room Right Pot 3",                ("Pot",          0x08,  (0,0,45), None,                        'Bombs (5)',                             ("Bottom of the Well MQ", "Master Quest", "Pots",))),
-   #("Bottom of the Well MQ Perimeter Behind Gate Pot",              ("Pot",          0x08,  0x2A, None,                            'N/A',                                   ("Bottom of the Well MQ", "Master Quest", "Pots",))),
+    ("Bottom of the Well MQ Perimiter Behind Gate Pot",              ("Pot",          0x08,  (0,0,47), None,                        'Fairy Drop',                            ("Bottom of the Well MQ", "Master Quest", "Pots",))),
```
```suggestion
    ("Bottom of the Well MQ Perimeter Behind Gate Pot",              ("Pot",          0x08,  (0,0,47), None,                        'Fairy Drop',                            ("Bottom of the Well MQ", "Master Quest", "Pots",))),
```

fenhl commented on SettingsList.py:
```diff
@@ -2613,8 +2610,22 @@ class SettingInfos:
+        }
+    )
+    shuffle_empty_pots = Checkbutton(
+        gui_text       = 'Include empty pots',
```
```suggestion
        gui_text       = 'Include Empty Pots',
```

fenhl commented on SettingsList.py:
```diff
@@ -2638,6 +2649,22 @@ class SettingInfos:
+    )
+
+    shuffle_empty_crates = Checkbutton(
+        gui_text       = 'Include empty crates',
```
```suggestion
        gui_text       = 'Include Empty Crates',
```

fenhl commented on new scene flags.txt:
This should go in the Notes directory.

fenhl commented on data/World/Bottom of the Well MQ.json:
```diff
@@ -28,6 +28,7 @@
             "Bottom of the Well MQ GS Basement": "can_child_attack",
             "Bottom of the Well MQ GS Coffin Room": "
                 (Small_Key_Bottom_of_the_Well, 2) and can_child_attack",
+            "Bottom of the Well MQ Perimiter Behind Gate Pot": "has_explosives and Slingshot",
```
```suggestion
            "Bottom of the Well MQ Perimeter Behind Gate Pot": "has_explosives and Slingshot",
```

fenhl commented on data/World/Overworld.json:
```diff
@@ -694,6 +694,25 @@
+            "GF Crate 4": "can_break_crate",
+            "GF Crate 5": "can_break_crate",
+            "GF Crate 6": "can_break_crate",
+            "GF HBA Crate 1":  "((is_adult and Gerudo_Membership_Card) or is_child) and can_break_crate",
```
Maybe the archery range should be a separate region?

r0bd0g replied:
There are maybe a couple other cases along these lines but I figured, it wasn't broke. If these kinds of things are still there next time I go through it all, I'll clean up some things with extra regions at that time, whenever that happens to be.

rrealmuto committed:
New flag system

rrealmuto committed:
Add hack to set flags/chest types when spawning any actor

rrealmuto commented:
Now that wonderitems is merged I've rebased this and will start incorporating the changes identified here

rrealmuto committed:
Fix BOTW MQ Pot Typo

Co-authored-by: Fenhl <fenhl@fenhl.net>

rrealmuto committed:
Empty Pots/Crates GUI format

Co-authored-by: Fenhl <fenhl@fenhl.net>

rrealmuto committed:
Fix Nothing/Fairy Drop starting items

rrealmuto committed:
Pass CI checks

rrealmuto committed:
Fix flag check for goron pot, rupee circles, shadow pot

rrealmuto committed:
Small improvement to how flag data is stored/retrieved. Fix subflag actors not spawning the correct items, and not working when some of the subflags were disabled

fenhl commented:
There's a bug on this branch where if you break one of the pots in room 1 of Dodongo's Cavern (the baby dodongo corridor), the game freezes. Tested with this plando:

```json
{
    "settings": {
        "logic_rules": "none",
        "spawn_positions": ["child"],
        "dungeon_shortcuts_choice": "all",
        "starting_items": {
            "Deku Shield": 1,
            "Bomb Bag": 1
        }
    },
    "entrances": {
        "Child Spawn -> KF Links House": "Death Mountain"
    }
}
```

fenhl commented:
The freeze does not occur on Dev-Rob.

rrealmuto commented:
So the last commit here isn't up on my branch, and likely what caused the crash. I'll test tonight and depending on what's going on, might just revert that last commit.

fenhl commented:
Can confirm the freeze does not happen on commit 17ae3a5.

fenhl commented:
Another bug on commit 17ae3a5: A shuffled item in the pots in Dodongos Cavern Lower Lizalfos Pot 1 can't be picked up. Attempting to save the game after attempting to pick up the item breaks things to a point where soft/hard reset no longer work, and BizHawk crashes on core reboot. This bug also doesn't occur on Dev-Rob.

rrealmuto commented:
Next thing you know you'll pick up the item from the pot and get teleported to ganon 🤔

rrealmuto committed:
Fix alt_override_table size

rrealmuto commented:
That should hopefully fix it

fenhl commented:
Looks good! Maybe we could add a CI check to make sure this array is large enough?

rrealmuto commented:
Haha I have a check in the patcher, but it was checking the wrong size (which is currently hard-coded into patches.py, and I probably copied that block over from my branch). Perhaps that could be made to dynamically determine the size of the array from get_items.c or from asm_symbols.txt and included in CI

rrealmuto committed:
Fix alt flag calculation

rrealmuto committed:
Reorder alt flags for child kak

rrealmuto committed:
Update build tools to capture symbol lengths. Rework run time length checks to use the symbol lengths

rrealmuto committed:
Add CI check for xflag/alt override tables

rrealmuto committed:
Add override table to CI check. Clean up plando

rrealmuto commented:
> Looks good! Maybe we could add a CI check to make sure this array is large enough?

Done. I updated the format of data/generated/symbols.txt to include the size of data symbols which seems to work well. I updated the runtime tests to compare against these sizes, and added a check in CI.py as well.

rrealmuto committed:
Add table check for EXTENDED_OBJECT_TABLE

fenhl commented:
Another bug report: If crate shuffle is disabled, all crates drop green rupees. Tested in the Kakariko backyard where 3 of the crates are supposed to drop nothing and the 4th is supposed to drop a red rupee:

```json
{
    "settings": {
        "logic_rules": "none",
        "starting_age": "adult",
        "spawn_positions": ["adult"]
    },
    "entrances": {
        "Adult Spawn -> Temple of Time": "Kak Backyard"
    }
}
```

This bug also doesn't occur on Dev-Rob.

rrealmuto committed:
Subflag fixes

rrealmuto commented:
> Another bug report: If crate shuffle is disabled, all crates drop green rupees. Tested in the Kakariko backyard where 3 of the crates are supposed to drop nothing and the 4th is supposed to drop a red rupee:
> 
> ```json
> {
>     "settings": {
>         "logic_rules": "none",
>         "starting_age": "adult",
>         "spawn_positions": ["adult"]
>     },
>     "entrances": {
>         "Adult Spawn -> Temple of Time": "Kak Backyard"
>     }
> }
> ```
> 
> This bug also doesn't occur on Dev-Rob.

Should be fixed now.

fenhl commented on SettingsList.py:
```diff
@@ -2645,6 +2656,22 @@ class SettingInfos:
+        default        = False,
+        gui_tooltip    = '''\
+            Enabling this will include empty crates into the location
+            pool based on the Shuffle Pots setting chosen.
```
```suggestion
            pool based on the Shuffle Crates setting chosen.
```

rrealmuto committed:
Improve logic for DMC Pots with empty pots shuffled

r0bd0g commented:
I didn't realize the shuffle empty pots setting would put the items into the dmc pots as child...
Are the day crates in market still separate checks from the night crates?

rrealmuto commented:
Yes

fenhl commented on HintList.py:
```diff
@@ -524,6 +524,8 @@ def tokens_required_by_settings(world: World) -> int:
     'Ocarina C down Button':                                    (["a low note"], "the Ocarina C down Button", 'item'),
     'Ocarina C left Button':                                    (["a somewhat high note"], "the Ocarina C left Button", 'item'),
     'Ocarina C right Button':                                   (["a middle note"], "the Ocarina C right Button", 'item'),
+    'Fairy Drop':                                               (["an annoying companion"], "a Stray Fairy", 'item'),
```
We should make sure the cryptic hints make sense between bottled and stray fairies. I think “Navi's cousin” sounds more like a stray fairy than a bottled one, so I suggest adding it here and removing it from the `Bottle with Fairy` hints.

```suggestion
    'Fairy Drop':                                               (["an annoying companion", "Navi's cousin"], "a Stray Fairy", 'item'),
```

fenhl replied:
Remember to also remove this hint from the bottled fairy item.

rrealmuto committed:
Update HintList.py

Co-authored-by: Fenhl <fenhl@fenhl.net>

rrealmuto committed:
Merge branch 'upstream' into new_flags

rrealmuto commented:
I just merged with the (almost) latest main Dev. Probably just needs a bit of testing to make sure I didn't overlook anything during the merge. Main conflicts were just gi_draw IDs

rrealmuto commented:
Nothing seems broken after the testing I've done + other folks that have been playing on my branch since I merged. I'm sure I'll eat those words right after it gets merged

rrealmuto committed:
Update SettingsList.py

Co-authored-by: Fenhl <fenhl@fenhl.net>

fenhl commented on data/World/Overworld.json:
```diff
@@ -2303,10 +2359,10 @@
-            "DMC Near GC Pot 2": "is_adult",
-            "DMC Near GC Pot 3": "is_adult",
-            "DMC Near GC Pot 4": "is_adult"
+            "DMC Near GC Pot 1": "is_adult or shuffle_empty_pots",
+            "DMC Near GC Pot 2": "is_adult or shuffle_empty_pots",
+            "DMC Near GC Pot 3": "is_adult or shuffle_empty_pots",
+            "DMC Near GC Pot 4": "is_adult or shuffle_empty_pots"
```
What's the reason for writing the logic like this? Shouldn't it just be `True`?

fenhl replied:
Discussed with @rrealmuto in DMs, the reason is that without `shuffle_empty_pots`, these pots are empty as child but have shuffled items as adult.

fenhl committed:
Code review for potsanity 3.0

fenhl reviewed (changes requested):
Posted a review in PR form at https://github.com/rrealmuto/OoT-Randomizer/pull/107

rrealmuto committed:
Merge pull request #107 from fenhl/pots3-review

Code review for potsanity 3.0

rrealmuto committed:
Fix C compiler warnings

fenhl commented:
This is failing the `test_build_room_xflags` test after resolving merge conflicts.

rrealmuto commented:
weird I thought I fixed that unit test the other day. It had been broken for a while

rrealmuto commented:
hmm I seem to not have committed it 😅

rrealmuto committed:
Fix test_build_room_xflags unit test
